File: ledger/test/cases/heartbeat.py
# The MIT License (MIT)
#
# Copyright (c) 2021 RSK Labs Ltd
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies
# of the Software, and to permit persons to whom the Software is furnished to do
# so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

from .case import TestCase, TestCaseError
import secp256k1 as ec
import hmac
import hashlib


class Heartbeat(TestCase):
    EXPECTED_HEADER = "HSM:SIGNER:HB:4.0:"
    EHL = len(EXPECTED_HEADER)

    @classmethod
    def op_name(cls):
        return "heartbeat"

    def __init__(self, spec):
        self.ud_value = spec["udValue"]

        super().__init__(spec)

    def run(self, dongle, debug, run_args):
        def normalize_hex_sig_component(h):
            h = bytes.fromhex(h)
            h = int.from_bytes(h, byteorder="big", signed=False)
            return h.to_bytes(32, byteorder="big", signed=False)

        def get_message_part(msg, part):
            try:
                return ({
                    "best_block": lambda m: m[self.EHL*2:(self.EHL+32)*2],
                    "last_tx": lambda m: m[(self.EHL+32)*2:(self.EHL+32+8)*2],
                })[part](msg)
            except KeyError:
                raise TestCaseError(f"Unknown message part \"{part}\"")
            except Exception as e:
                raise TestCaseError(str(e))

        try:
            heartbeat = dongle.get_signer_heartbeat(self.ud_value)
            debug(f"Heartbeat: {heartbeat}")

            if not heartbeat[0]:
                error_code = heartbeat[1]
                raise TestCaseError("Expected success getting the heartbeat "
                                    f"but got error code {error_code}")
            heartbeat = heartbeat[1]

            # Validate signature
            message = bytes.fromhex(heartbeat["message"])
            pubkey = ec.PublicKey(bytes.fromhex(heartbeat["pubKey"]), raw=True)
            norm_r = normalize_hex_sig_component(heartbeat["signature"].r)
            norm_s = normalize_hex_sig_component(heartbeat["signature"].s)
            sig = pubkey.ecdsa_deserialize_compact(norm_r + norm_s)
            tweak = hmac.new(
                    bytes.fromhex(heartbeat["tweak"]),
                    pubkey.serialize(compressed=False),
                    hashlib.sha256,
                ).digest()
            pubkey = pubkey.tweak_add(tweak)

            if not pubkey.ecdsa_verify(message, sig):
                raise TestCaseError("Expected signature to be valid but it wasn't")

            # Validate header
            header_msg = message[:self.EHL].decode('ascii')
            if header_msg != self.EXPECTED_HEADER:
                raise TestCaseError(f"Expected header to be {self.EXPECTED_HEADER} but"
                                    f" got {header_msg}")

            # Validate UD value
            ud_msg = message[-16:].hex()
            if ud_msg != self.ud_value:
                raise TestCaseError(f"Expected UD value to be {self.ud_value} but"
                                    f" got {ud_msg}")

            # Expectations on the heartbeat message (optional)
            if type(self.expected) == dict:
                message = message.hex()
                for key in self.expected:
                    val = get_message_part(message, key)
                    if val != self.expected[key]:
                        raise TestCaseError(f"Expected {key} to be {self.expected[key]} "
                                            f"but got {val}")
        except RuntimeError as e:
            raise TestCaseError(str(e))

File: ledger/src/signer/src/heartbeat.c
/**
 * The MIT License (MIT)
 *
 * Copyright (c) 2021 RSK Labs Ltd
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to
 * deal in the Software without restriction, including without limitation the
 * rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
 * sell copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
 * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
 * IN THE SOFTWARE.
 */

#include <string.h>

#include "os.h"
#include "heartbeat.h"
#include "apdu.h"
#include "defs.h"
#include "memutil.h"
#include "bc_state.h"
#include "compiletime.h"

/*
 * Reset the given heartbeat context
 */
static void reset_heartbeat(heartbeat_t* heartbeat_ctx) {
    explicit_bzero(heartbeat_ctx, sizeof(heartbeat_t));
    heartbeat_ctx->state = STATE_HEARTBEAT_WAIT_UD_VALUE;
}

/*
 * Check that the SM for the heartbeat generation
 * matches the expected state.
 *
 * Reset the state and throw a protocol error
 * otherwise.
 */
static void check_state(heartbeat_t* heartbeat_ctx,
                        state_heartbeat_t expected) {
    if (heartbeat_ctx->state != expected) {
        reset_heartbeat(heartbeat_ctx);
        THROW(ERR_HBT_PROT_INVALID);
    }
}

// -----------------------------------------------------------------------
// Protocol implementation
// -----------------------------------------------------------------------

/*
 * Implement the heartbeat protocol.
 *
 * @arg[in] rx      number of received bytes from the Host
 * @arg[in] heartbeat_ctx heartbeat context
 * @ret             number of transmited bytes to the host
 */
unsigned int get_heartbeat(volatile unsigned int rx,
                           heartbeat_t* heartbeat_ctx) {
    // Build should fail when more bytes than available are tried to be copied
    // from the last signed BTC tx hash
    COMPILE_TIME_ASSERT(LAST_SIGNED_TX_BYTES <=
                        sizeof(N_bc_state.last_auth_signed_btc_tx_hash));

    COMPILE_TIME_ASSERT(MAX_HEARTBEAT_MESSAGE_SIZE <= APDU_TOTAL_DATA_SIZE_OUT);

    switch (APDU_OP()) {
    case OP_HBT_UD_VALUE:
        // Should receive a user-defined value
        if (APDU_DATA_SIZE(rx) != UD_VALUE_SIZE) {
            reset_heartbeat(heartbeat_ctx);
            THROW(ERR_HBT_PROT_INVALID);
        }

        // Initialize message
        explicit_bzero(heartbeat_ctx->msg, sizeof(heartbeat_ctx->msg));
        heartbeat_ctx->msg_offset = 0;

        // Copy the message prefix
        SAFE_MEMMOVE(heartbeat_ctx->msg,
                     sizeof(heartbeat_ctx->msg),
                     heartbeat_ctx->msg_offset,
                     (void*)PIC(HEARTBEAT_MSG_PREFIX),
                     HEARTBEAT_MSG_PREFIX_LENGTH,
                     MEMMOVE_ZERO_OFFSET,
                     HEARTBEAT_MSG_PREFIX_LENGTH,
                     THROW(ERR_HBT_INTERNAL));
        heartbeat_ctx->msg_offset += HEARTBEAT_MSG_PREFIX_LENGTH;

        // Copy the current best block
        SAFE_MEMMOVE(heartbeat_ctx->msg,
                     sizeof(heartbeat_ctx->msg),
                     heartbeat_ctx->msg_offset,
                     N_bc_state.best_block,
                     sizeof(N_bc_state.best_block),
                     MEMMOVE_ZERO_OFFSET,
                     sizeof(N_bc_state.best_block),
                     THROW(ERR_HBT_INTERNAL));
        heartbeat_ctx->msg_offset += sizeof(N_bc_state.best_block);

        // Copy the last LAST_SIGNED_TX_BYTES bytes of the last auth signed tx
        SAFE_MEMMOVE(heartbeat_ctx->msg,
                     sizeof(heartbeat_ctx->msg),
                     heartbeat_ctx->msg_offset,
                     N_bc_state.last_auth_signed_btc_tx_hash,
                     sizeof(N_bc_state.last_auth_signed_btc_tx_hash),
                     MEMMOVE_ZERO_OFFSET,
                     LAST_SIGNED_TX_BYTES,
                     THROW(ERR_HBT_INTERNAL));
        heartbeat_ctx->msg_offset += LAST_SIGNED_TX_BYTES;

        // Copy the UD value from the APDU
        SAFE_MEMMOVE(heartbeat_ctx->msg,
                     sizeof(heartbeat_ctx->msg),
                     heartbeat_ctx->msg_offset,
                     APDU_DATA_PTR,
                     APDU_TOTAL_DATA_SIZE,
                     MEMMOVE_ZERO_OFFSET,
                     UD_VALUE_SIZE,
                     THROW(ERR_HBT_INTERNAL));
        heartbeat_ctx->msg_offset += UD_VALUE_SIZE;

        heartbeat_ctx->state = STATE_HEARTBEAT_READY;

        return TX_FOR_DATA_SIZE(0);
    case OP_HBT_GET:
        check_state(heartbeat_ctx, STATE_HEARTBEAT_READY);

        // Sign message
        int endorsement_size = os_endorsement_key2_derive_sign_data(
            heartbeat_ctx->msg, heartbeat_ctx->msg_offset, APDU_DATA_PTR);

        return TX_FOR_DATA_SIZE(endorsement_size);
    case OP_HBT_GET_MESSAGE:
        check_state(heartbeat_ctx, STATE_HEARTBEAT_READY);

        SAFE_MEMMOVE(APDU_DATA_PTR,
                     APDU_TOTAL_DATA_SIZE,
                     MEMMOVE_ZERO_OFFSET,
                     heartbeat_ctx->msg,
                     sizeof(heartbeat_ctx->msg),
                     MEMMOVE_ZERO_OFFSET,
                     heartbeat_ctx->msg_offset,
                     THROW(ERR_HBT_INTERNAL));

        return TX_FOR_DATA_SIZE(heartbeat_ctx->msg_offset);
    case OP_HBT_APP_HASH:
        return TX_FOR_DATA_SIZE(os_endorsement_get_code_hash(APDU_DATA_PTR));
    case OP_HBT_PUBKEY:
        return TX_FOR_DATA_SIZE(os_endorsement_get_public_key(
            ENDORSEMENT_SCHEME_INDEX, APDU_DATA_PTR));
    default:
        reset_heartbeat(heartbeat_ctx);
        THROW(ERR_HBT_PROT_INVALID);
        break;
    }
}
